### ISSUE TITLE
Clustering metrics

### ISSUE THREAD

**Issue by SkafteNicki** [Lightning-AI/torchmetrics#2003]
## 🚀 Feature

Lets add clustering metrics to TM:
- [x] [Dunn index](https://en.wikipedia.org/wiki/Dunn_index) -> #2049 
- [ ] [Silhouette](https://scikit-learn.org/stable/modules/generated/sklearn.metrics.silhouette_score.html) @stancld 
- [x] [Rand score](https://scikit-learn.org/stable/modules/generated/sklearn.metrics.rand_score.html#sklearn.metrics.rand_score) -> #2025
- [x] [Adjusted_mutual_info_score](https://scikit-learn.org/stable/modules/generated/sklearn.metrics.adjusted_mutual_info_score.html#sklearn.metrics.adjusted_mutual_info_score) -> #2058
- [x] [Adjusted rand score](https://scikit-learn.org/stable/modules/generated/sklearn.metrics.adjusted_rand_score.html#sklearn.metrics.adjusted_rand_score) -> #2032 
- [x] [Calinski Harabasz score](https://scikit-learn.org/stable/modules/generated/sklearn.metrics.calinski_harabasz_score.html#sklearn.metrics.calinski_harabasz_score) -> #2036 
- [x] [Completeness score](https://scikit-learn.org/stable/modules/generated/sklearn.metrics.completeness_score.html#sklearn.metrics.completeness_score) -> #2053
- [x] [Fowlkes mallows score](https://scikit-learn.org/stable/modules/generated/sklearn.metrics.fowlkes_mallows_score.html#sklearn.metrics.fowlkes_mallows_score) -> #2066 
- [x] [Homogeneity completeness measure](https://scikit-learn.org/stable/modules/generated/sklearn.metrics.homogeneity_completeness_v_measure.html#sklearn.metrics.homogeneity_completeness_v_measure) -> #2053
- [x] [Mutual info score](https://scikit-learn.org/stable/modules/generated/sklearn.metrics.mutual_info_score.html#sklearn.metrics.mutual_info_score) -> #2008 
- [x] [Normalized mutual info score](https://scikit-learn.org/stable/modules/generated/sklearn.metrics.normalized_mutual_info_score.html#sklearn.metrics.normalized_mutual_info_score) -> #2029 
- [x] [Davies–Bouldin index](https://en.wikipedia.org/wiki/Davies%E2%80%93Bouldin_index) -> #2071
- [x] [V measure score](https://scikit-learn.org/stable/modules/generated/sklearn.metrics.v_measure_score.html#sklearn.metrics.v_measure_score) -> #2053

### Motivation

In Supervised Learning, the labels are known and evaluation can be done by calculating the degree of correctness by comparing the predicted values against the labels. However, in Unsupervised Learning, the labels are not known, which makes it hard to evaluate the degree of correctness as there is no ground truth.

That being said, it is still consistent that a good clustering algorithm has clusters that have small within-cluster variance (data points in a cluster are similar to each other) and large between-cluster variance (clusters are dissimilar to other clusters).

ref: https://towardsdatascience.com/7-evaluation-metrics-for-clustering-algorithms-bdc537ff54d2

### CTA

pls also check our contribution guide: https://torchmetrics.readthedocs.io/en/stable/generated/CONTRIBUTING.html

**Comment by matsumotosan** [comment]
Hi @SkafteNicki, I'd like to give Mutual Info Score a shot!

**Comment by SkafteNicki** [comment]
@matsumotosan cool thanks for wanting to contribute :]
Yes, please add new directories as I consider this a new domain of metrics.

**Comment by SkafteNicki** [comment]
Also if it help, I was fooling around with some code in the weekend, trying to calculate the contingency matrix and pair matrix which seems to be two common objects used for calculating clustering metrics:
```python
p = torch.tensor([0, 0, 1, 2])
t = torch.tensor([0, 0, 1, 1])

p_u = len(torch.unique(p))
t_u = len(torch.unique(t))

lin = p + p_u * t
contingency_matrix = torch.bincount(t + t_u * p, minlength=p_u * t_u).reshape(p_u, t_u)

n_c = contingency_matrix.sum(dim=1)
n_k = contingency_matrix.sum(dim=0)
sum_squared = (contingency_matrix ** 2).sum()

pair_matrix = torch.zeros(2,2)
pair_matrix[1,1] = sum_squared - len(p)
pair_matrix[0,1] = (contingency_matrix * n_k).sum() - sum_squared
pair_matrix[1,0] = (contingency_matrix.T * n_c).sum() - sum_squared
pair_matrix[0,0] = len(p) ** 2 - pair_matrix[0,1] - pair_matrix[1,0] - sum_squared
```
probably needs to be improved but maybe useful.

**Comment by matsumotosan** [comment]
I [tried](https://github.com/Lightning-AI/torchmetrics/pull/2008) calculating the contingency matrix based on sklearn's [implementation](https://github.com/scikit-learn/scikit-learn/blob/7f9bad99d6e0a3e8ddf92a7e5561245224dab102/sklearn/metrics/cluster/_supervised.py#L154C1-L158C6). I'll compare it to the method you suggested since it does seem to be a key computation.

**Comment by matsumotosan** [comment]
I'd like to work on normalized mutual info and Dunn index next!

**Comment by SkafteNicki** [comment]
> I'd like to work on normalized mutual info and Dunn index next!

Assigned :]

**Comment by matsumotosan** [comment]
@SkafteNicki Would it be useful if we accumulated all the inputs for clustering-related tests in a separate file similar to `inputs.py` for classification? I saw that your `rand_score` tests used the same test cases that I used for `mutual_info` and `normalized_mutual_info`. I could also add cases for Dunn index that may be useful for other metrics in the future.

**Comment by stancld** [comment]
@SkafteNicki I'll take Silhouette

**Comment by SkafteNicki** [comment]
> @SkafteNicki Would it be useful if we accumulated all the inputs for clustering-related tests in a separate file similar to `inputs.py` for classification? I saw that your `rand_score` tests used the same test cases that I used for `mutual_info` and `normalized_mutual_info`. I could also add cases for Dunn index that may be useful for other metrics in the future.

@matsumotosan I think that is a good idea, anything to help reduce replicated code is always good. 
Can you send a PR with this ? Else I do it

**Comment by matsumotosan** [comment]
I'll send a draft PR in a bit.

**Comment by matsumotosan** [comment]
@SkafteNicki I'll take adjusted mutual info score as well.

**Comment by m0saan** [comment]
Hello @SkafteNicki, which of these are not yet claimed?

**Comment by SkafteNicki** [comment]
> Hello @SkafteNicki, which of these are not yet claimed?

Hi @m0saan,
Currently [Fowlkes mallows score](https://scikit-learn.org/stable/modules/generated/sklearn.metrics.fowlkes_mallows_score.html#sklearn.metrics.fowlkes_mallows_score) and [Davies–Bouldin index](https://en.wikipedia.org/wiki/Davies%E2%80%93Bouldin_index) is not yet claimed.

**Comment by matsumotosan** [comment]
@SkafteNicki I can take Fowlkes-Mallows.

**Comment by Borda** [comment]
Most have been done, Thank you all for your contributions! :chipmunk:
